File: docs/JAMIA/comparisons/RPython/TM/12125364_median_int_vs_double.ts
/**
 * RPython SO#12125364 — Why does median trip up data.table (integer vs double)?
 * Effect: Crash
 * Bug class: Int/double distinction
 *
 * In R data.table, median() returns integer for odd-length groups and double for
 * even-length groups. data.table requires consistent types across groups, causing
 * a crash.
 *
 * In JavaScript, there is no int/double distinction — all numbers are `number`.
 * This class of bug does not exist. The type system still catches unrelated
 * misuse (e.g., passing a string column to s.median()).
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const data = createDataFrame([
  { day: "Mon", n: 3 },
  { day: "Mon", n: 5 },
  { day: "Mon", n: 2 },
  { day: "Tue", n: 7 },
  { day: "Tue", n: 4 },
  { day: "Wed", n: 6 },
  { day: "Wed", n: 1 },
  { day: "Wed", n: 8 },
]);

// s.median() returns number regardless of group size — no int/double conflict
const result = data.groupBy("day").summarize({
  patient_encounters: (g) => s.median(g.extract("n")),
});

// Passing a string column to median is still caught
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.median(data.extract("day"));

result.print();

File: docs/JAMIA/comparisons/RPython/TM/12125364_median_int_vs_double.R
# ID: SO#12125364
# Language: R
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: median() returns int for odd-length groups, double for even-length. data.table crashes on inconsistent return types across groups. Not verified with R runtime — .R file written from SO code.
# Reproduction status: Fixed (data.table later enforces consistent return types)
# Type system catch: Grouped aggregation enforces consistent return type across groups

library(data.table)

set.seed(42)
enc.per.day <- data.table(
  DOS = as.Date("2023-01-01") + 0:13,
  n = as.integer(c(3, 5, 2, 7, 4, 6, 1, 8, 3, 5, 2, 7, 4, 6))
)

# median of odd-length group returns integer, even-length returns double
# data.table requires consistent types across groups → crash
enc.per.day[, list(patient.encounters = median(n)), by = list(weekdays(DOS))]
# Error: Column 1 of result for group X is type 'double' but expecting type 'integer'
